src/Components/SourcesDropdown/index.js

```diff
@@ -1,58 +1,65 @@
-import React from 'react';
-import { Dropdown, DropdownToggle, DropdownItem } from '@patternfly/react-core';
-import { provisioningUrl } from '../../API/helpers';
+import React from "react";
+import {
+  Dropdown,
+  DropdownToggle,
+  DropdownItem,
+  Spinner,
+} from "@patternfly/react-core";
+import { useQuery } from 'react-query';
+
+import withQueryClient from "../Common/Query";
+import { SOURCES_QUERY_KEY } from "../../API/consts";
+import { fetchSourcesList } from "../../API";
 
 const SourcesDropdown = () => {
   const [isOpen, setIsOpen] = React.useState(false);
-  const [sources, setSources] = React.useState([]);
+  const [chosenSource, setChosenSource] = React.useState();
+
+  const {
+    isLoading,
+    error,
+    data: sources,
+  } = useQuery(SOURCES_QUERY_KEY, fetchSourcesList);
+
   const onToggle = (isOpen) => {
-    fetchSources();
     setIsOpen(isOpen);
   };
   const onFocus = () => {
-    const element = document.getElementById('toggle-secondary');
+    const element = document.getElementById("toggle-secondary");
     element.focus();
   };
   const onSelect = () => {
     setIsOpen(false);
     onFocus();
   };
 
-  //TODO: move to react-query
-  const fetchSources = async () => {
-    const response = await fetch(provisioningUrl('sources'));
-    const json = await response.json();
-    setSources(json);
-  };
-
   // TODO: add the chosen source to an global/external state
-  const sourcesDetails = () => console.log(`${name} was chosen`);
+  const sourcesDetails = id => {
+    console.log(`${id} was chosen`);
+    setChosenSource(id)
+  }
+  const dropdownItemsMapper = () => {
+    if (isLoading)
+      return (
+        <DropdownItem>
+          <Spinner size="md" isSVG />
+        </DropdownItem>
+      );
+    if (sources?.length > 0) {
```

Comment:
why is there a '?'

Reply:
represents [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) :)
if `sources` is null, then `sources.length` will produce a runtime error
the optional chaining stops chaining if theres a null